Fixed #22288 -- Fixed F() expressions with the  __range lookup.

MatthewWilkes opened the pull request:
This pull request intends to clarify some handling of lookups that take a tuple of arguments, like Range and In, and allow them to contain F() objects or expressions as arguments.

https://code.djangoproject.com/ticket/22288

jarshwah commented:
Hi Matthew. I've only had a quick look over this so far (sorry!), but the approach you've taken looks pretty good. You're accounting for future lookups that take iterables which is a plus. A couple of quick notes:

- `MultipleArgumentLookup.batch_process_rhs` took me longer to figure out what was happening than I should probably admit to :) perhaps a few comments would be good.

- The changes in `build_filter` seem excessive but they might be necessary. If there's any way to push that logic into `lookups` that would be better I think.

I need to put some more work into reviewing the change properly, just wanted to let you know that you're definitely on the right track. @akaariai any chance you can lend your experience to review?

akaariai commented on django/db/models/sql/query.py:
```diff
@@ -1137,7 +1137,29 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
+            klass = type(value)
+            new_value = []
+            new_lookups = []
+            new_used_joins = []
```
Make this and new_lookups sets.

MatthewWilkes replied:
I tried making new_lookups a set but this caused non-deterministic failures when running the tests under postgres, due to the lookups being interpreted as transforms when they're returned out of order.

akaariai commented on django/db/models/sql/query.py:
```diff
@@ -1137,7 +1137,29 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
         # Work out the lookup type and remove it from the end of 'parts',
         # if necessary.
-        value, lookups, used_joins = self.prepare_lookup_value(value, lookups, can_reuse, allow_joins)
+        if isinstance(value, (list, tuple)) and value:
```
The check should be inside prepare_lookup_value, not in build_filter.

akaariai commented on django/db/models/sql/query.py:
```diff
@@ -1137,7 +1137,29 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
+                new_value.append(inner_value)
+                for lookup in lookups:
+                    if lookup not in new_lookups:
+                        new_lookups.append(lookup)
```
I believe this to be broken. It shouldn't be possible for one of the rhs values to produce a new lookup which others didn't produce.

akaariai commented on django/db/models/sql/query.py:
```diff
@@ -1137,7 +1137,29 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
+                        new_lookups.append(lookup)
+                for used_join in used_joins:
+                    if used_join not in new_used_joins:
+                        new_used_joins.append(used_join)
```
For joins, we might end up doing LEFT OUTER JOIN for cases where INNER JOIN would be optimal. I don't think that is a blocker. But if we do a INNER JOIN for a case where we should be using LEFT OUTER JOIN, then the query will produce wrong results.

For testing this models structure like
```
class Timestamp(models.Model):
    at_time = models.DateTimeField()

class Event(models.Model):
    start = models.ForeignKey(Timestamp, null=True)
    end = models.ForeignKey(Timestamp, null=True)
```
should be tested with range queries, where one or both of the start and end time are null.

MatthewWilkes replied:
Hi @akaariai,

I'm wondering if you could clarify this for me please? What about the aggregation of required joins from the independent from the sub-values would cause the join type to change? It seems to just be table names without metadata.

Matt

jarshwah replied:
I think what Anssi is asking you to do is prove that the following produces LEFT OUTER joins rather than INNER JOINs.

```
class Timestamp(models.Model):
    at_time = models.DateTimeField()

class Event(models.Model):
    start = models.ForeignKey(Timestamp, null=True)
    end = models.ForeignKey(Timestamp, null=True)
    actual = models.DateTimeField()

start_datetime = datetime.datetime(2016, 6, 2)
end_datetime = datetime.datetime(2016, 6, 4)
actual = datetime.datetime(2016, 6, 3)
t1 = Timestamp.objects.create(at_time=start_datetime)
t2 = Timestamp.objects.create(at_time=end_datetime)

Event.objects.create(start=t1, end=t2, actual=actual)
Event.objects.create(start=t1, end=None, actual=actual)
Event.objects.create(start=None, end=t2, actual=actual)
Event.objects.create(start=None, end=None, actual=actual)

# this should produce LEFT OUTER JOINS, not INNER JOIN
Event.objects.filter(actual__range=[F('start__at_time'), F('end__at_time')])

```

I'm not sure what those results will be, but the joins should be LEFT OUTER.

akaariai commented on django/db/models/sql/query.py:
```diff
@@ -1137,7 +1137,29 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
+                        new_used_joins.append(used_join)
+            # Tuples are handled differently to lists when building clauses,
+            # so make sure we retain the type we originally received
+            value, lookups, used_joins = klass(new_value), new_lookups, new_used_joins
```
What happens with foofield__exact=(1, None). What happens if foofield is an ArrayField from contrib.postgres?

MatthewWilkes replied:
I've added tests for ArrayFields, but I'm going to have to declare defeat for at least a subset of the problem. Allowing F objects and expressions to be members of the array being searched for is causing me problems in the SQL generation, as mogrify is more complex for Array than simple field lookups.

I have F objects that refer to an array type themselves working, which I think is good enough (tm).

akaariai commented:
While I believe the approach in this patch is workable with some tuning, I'd much prefer moving all this logic into lookups. We can do that now as the old style lookups are removed.

So, we could first resolve the lookup class to be used in the query, then ask that lookup class to prepare_lookup_value(query, lookups, can_reuse, allow_joins). Now the lookup could decide what exactly to do with list/tuple as rhs.

In general build_filter() would benefit from moving logic to the Lookup class.

I'm not sure if we have to do the move of prepare_lookup_value() before adding support for multipart rhs. My initial hunch is that it is nearly as easy to do the move after this PR, so we can proceed with this PR and do the refactoring later on.

MatthewWilkes commented:
Thanks for the comments, @jarshwah and @akaariai. I also wondered about how much to offload onto the lookups, but seems I only mentioned that in the trac ticket. Sorry about that.

I was a little nervous about this moving from a bug fix into a bigger refactoring of how the parsing code works, especially as my understanding of it is limited.

You've certainly unblocked me from knowing what my next steps are, thank you!

MatthewWilkes commented:
FYI: I am going to go for the more in-depth restructuring, I think. I hope to put something together in the coming week.

MatthewWilkes commented:
@timgraham @claudep  It looks like #6499 obsoletes the improvements that were WIP here, do you agree? If so, does it make sense for me to make this a tests-only contribution?

claudep commented:
+1 for keeping the tests

timgraham commented on tests/expressions/tests.py:
```diff
@@ -331,6 +331,101 @@ def test_object_update_unsaved_objects(self):
+        Company.objects.filter(num_employees__exact=F("num_chairs"))
+
+        queryset = Company.objects.filter(num_employees__range=(F('num_chairs'), 100))
+        self.assertQuerysetEqual(queryset, ['<Company: Test GmbH>',
```
When updating the tests, please use the indentation style of the existing tests, e.g.
```python
self.assertQuerysetEqual(
    Number.objects.all(),
    [
        '<Number: -1, -1.000>',
        '<Number: 43, 42.000>',
        '<Number: 1338, 1337.000>'
    ],
    ordered=False
)
```

timgraham commented:
#6499 is merged but the new tests don't pass on master so I guess code changes are still needed.

MatthewWilkes commented:
If only I were at Pycon and looking for something to do for the sprints.

Oh… wait… oh…

Thanks for the heads up @timgraham, I admit I'd lost enthusiasm because I thought it had already been solved.

jarshwah commented:
buildbot, add to whitelist.

jarshwah commented:
buildbot, test this please.

jarshwah commented:
buildbot, test on oracle.

MatthewWilkes commented:
Looks like there are some legitimate failures due to F expressions not being datetimes, when it comes to detecting timezones. Also, there appear to be a couple of bogus ones from a change today to the master branch, around redirects.

MatthewWilkes commented:
buildbot, test on oracle.

MatthewWilkes commented:
No idea if I can trigger that, of course.

carljm commented on django/db/backends/mysql/operations.py:
```diff
@@ -142,6 +143,8 @@ def adapt_datetimefield_value(self, value):
             return None
 
         # MySQL doesn't support tz-aware datetimes
+        if isinstance(value, Combinable):
```
It looks like this new clause (and similar ones below) are splitting a comment from the code it applies to.

MatthewWilkes replied:
I think you might have a point. My logic for putting it there was that that check for time zone is the only reason the short circuit is required, as combinables aren't date times themselves. Happy to move it though.

carljm replied:
If the immediately following clause were really the _only_ reason the short-circuit was required, then I'd say don't short-circuit, just include the combinable check in a compound conditional for that clause. That way you avoid the possibility of accidentally short-circuiting other code that might later be added later in the method, that shouldn't be short-circuited.

But: I don't think it's accurate that the following clause is the only thing that can't handle a Combinable -- the `supports_microsecond_precision` check following it also should be skipped. In general it seems likely to me that the remainder of this method will always assume it has a datetime, so a short-circuit return is correct. So I think the comment should stay with the clause it applies to.

carljm commented on tests/postgres_tests/test_array.py:
```diff
@@ -173,12 +173,32 @@ def test_in(self):
             self.objs[:2]
         )
 
+    @unittest.expectedFailure
```
Should this expected-failure decorator be removed?

MatthewWilkes replied:
These two tests test for another use case of this feature that I've been unable to implement.

Specifically, you can now provide a F object that refers to an array when doing filters on an array field, which is tested and working, however you cannot use an F object in constructing an array that goes into the filter.

I think it is better to demonstrate that this extension to the general feature doesn't work, but I'll defer to the core team's judgement on that.

carljm replied:
Got it. I agree that adding expectedFailure tests is a good approach in this case.

timgraham replied:
Please add comments explaining the rationale. (see https://code.djangoproject.com/ticket/23943 for past confusion about the purpose of `expectedFailure`)

MatthewWilkes replied:
Added.

timgraham replied:
Are you going to create a ticket for this issue we could reference in the comment?

MatthewWilkes replied:
Added #27095 and referenced in the comments.

carljm commented on tests/postgres_tests/test_array.py:
```diff
@@ -173,12 +173,32 @@ def test_in(self):
             self.objs[:2]
         )
 
+    @unittest.expectedFailure
```
and this one?

carljm commented on tests/expressions/tests.py:
```diff
@@ -332,12 +336,195 @@ def test_object_update_unsaved_objects(self):
         with self.assertRaisesMessage(ValueError, msg):
             acme.save()
 
+    def test_ticket_22288_in_f_integer(self):
```
Usually we prefer for test names to describe the functionality, not reference the ticket. git blame can always be used to get back to the ticket if needed.

MatthewWilkes replied:
I had been taking my lead from other functions in that class, like test_ticket_18375_chained_filters and test_ticket_18375_kwarg_ordering_2. I'll switch mine.

carljm commented on tests/expressions/tests.py:
```diff
@@ -332,12 +336,195 @@ def test_object_update_unsaved_objects(self):
+        # MySQL requires that the values calculated for expressions do not
+        # pass outside of the field's range. We delete the standard company
+        # objects as they have inconvenient values for these tests
+        Company.objects.all().delete()
```
Maybe just put these tests in a separate TestCase to avoid useless creation and deletion of companies? Especially since you have two tests doing identical setup of alternate company data.

MatthewWilkes replied:
I've done this.

MatthewWilkes commented:
I've addressed the comments on here, but I'm aware it may still need more of a thorough review from an ORM expert since my last refactoring.

MatthewWilkes commented:
@akaariai @jarshwah do you have any more thoughts on this, given it was reduced in scope by the other PR and refactored to address your comments?

It would be nice to get a yay/nay or some changes to be made before it grows conflicts again.

jarshwah commented on django/db/backends/oracle/operations.py:
```diff
@@ -399,6 +400,10 @@ def adapt_datetimefield_value(self, value):
             return None
 
+        # Expression values are handled by the database so are automatically correctly adapted
+        if isinstance(value, Combinable):
```
The "right" way to check if something is an expression is by writing `hasattr(value, 'resolve_expression')`. Not all expressions have to be combinables - but they all must be resolvable. There's a few other places this same check is done too, I won't comment on all of them.

MatthewWilkes replied:
I've addressed each instance.

jarshwah commented:
Sorry I've taken so long to respond. I haven't given it a thorough review yet (sorry!) but everything appears fine to me. I'll give it a clearer look tomorrow.

jarshwah commented on django/db/models/lookups.py:
```diff
@@ -190,6 +198,49 @@ class FieldGetDbPrepValueIterableMixin(FieldGetDbPrepValueMixin):
+
+    def process_rhs(self, compiler, connection):
+        if self.rhs_is_direct_value():
+            # rhs should be an iterable of 2 values, we use batch_process_rhs
```
Couldn't there be more than 2 values for `__in` queries, or is this code not triggered for `__in`?

MatthewWilkes replied:
It doesn't actually require 2 values. I've clarified this comment and added an assert to verify that in works with three values.

jarshwah commented on django/db/models/lookups.py:
```diff
@@ -190,6 +198,49 @@ class FieldGetDbPrepValueIterableMixin(FieldGetDbPrepValueMixin):
+    def batch_process_rhs(self, compiler, connection, rhs=None):
+        pre_processed = super(FieldGetDbPrepValueIterableMixin, self).batch_process_rhs(compiler, connection, rhs)
+        params = []
+        for sql, param in zip(*pre_processed):
```
A comment here would be nice, something like:

```
# Expressions can not be parameters so replace them with their SQL and add to the existing query.
```

MatthewWilkes replied:
I looked at this more closely, and realised I'd been a bit naïve with the implementation of resolve_combinable_parameter. I've now generalised it a bit to allow for more complex lookups and prevented a nasty sql injection vulnerability (and added a new test to cover that).

jarshwah commented:
The design here is OK I think. Lookups are responsible for more and new iterable lookups should now be handled correctly. There's a few notes I made above to fix and some release notes, and then we can give it a thorough style review.

MatthewWilkes commented:
Thanks, I'm on holiday in Italy with limited internet access until Sunday, I'll update and push early next week.

I'm *so* going to try and find something I need to fix in something simple next time..

MatthewWilkes commented:
buildbot, test on oracle.

MatthewWilkes commented:
buildbot, test on oracle.

MatthewWilkes commented:
@jarshwah I've made the relevant changes here.

I've added another test, which I feel is a bit janky, but I don't know how to do it better. The context is that part of the logic had survived from my earliest attempts to get this working, before I really understood what I was doing, and I had never revisited it. When I was looking over from your comments last time it was glaring.

I'm happy for the test to not make it in, or in another form, but thought it work committing for visibility.

jarshwah commented:
@MatthewWilkes what was the name of the new test that you feel is janky so I can give it a thorough look over. Just need a quick mention in the release notes and we can work at getting this merged. I don't think we need to add a note to the queryset docs because there's no existing mention there of F expressions.

MatthewWilkes commented:
@jarshwah Sorry, I forgot that it wouldn't be obvious because of the squashing. The nasty one is test_complex_expressions_do_not_introduce_sql_injection_via_untrusted_string_inclusion

I did look at the release notes previously and wasn't quite sure where to put it, a minor feature listed under models?

jarshwah commented on tests/expressions/tests.py:
```diff
@@ -391,6 +401,218 @@ def test_ticket_18375_chained_filters(self):
+        Testing against sqlite demonstrates that the problem doesn't exist while keeping the test simple as it's the
+        most permissive of the supported databases.
+        """
+        Company.objects.filter(num_employees__exact=F("num_chairs"))
```
This line left by accident?

MatthewWilkes replied:
Good spot, removed.

jarshwah commented:
Leave the test in, it's certainly something we'd want to know about if it starts failing. Good pickup! Release notes should go under models heading.

MatthewWilkes commented:
@jarshwah Okay, I've removed that stray line and added a release note in 1.11 under models.

jarshwah commented on django/db/models/lookups.py:
```diff
@@ -190,6 +198,53 @@ class FieldGetDbPrepValueIterableMixin(FieldGetDbPrepValueMixin):
+            return self.rhs._prepare(self.lhs.output_field)
+        for rhs_value in self.rhs:
+            if hasattr(rhs_value, 'resolve_expression'):
+                # A combinable expression will be handled by the database, but
```
`An expression` - no need for the `combinable` bit.

jarshwah commented on django/db/models/lookups.py:
```diff
@@ -190,6 +198,53 @@ class FieldGetDbPrepValueIterableMixin(FieldGetDbPrepValueMixin):
+        else:
+            return super(FieldGetDbPrepValueIterableMixin, self).process_rhs(compiler, connection)
+
+    def resolve_combinable_parameter(self, compiler, connection, sql, param):
```
`resolve_expression_parameter` maybe? You're not really dealing with combinables here (even though they are also combinable), so just go with expression based names I think.

jarshwah commented on django/db/models/lookups.py:
```diff
@@ -190,6 +198,53 @@ class FieldGetDbPrepValueIterableMixin(FieldGetDbPrepValueMixin):
+        # The TO_DELETE singleton represents parameters that have been integrated directly into sql, so they are
+        # replaced to avoid a mismatch of number of format strings and parameters.
+        params = [item for item in params if item is not TO_DELETE]
+        return sql, params
```
You can reduce iterations of params by unpacking in a single pass:

```
        params = []
        for sql, param in zip(*pre_processed):
            params.append(self.resolve_combinable_parameter(compiler, connection, sql, param))
        sql, params = zip(*params)
        # The TO_DELETE singleton represents parameters that have been integrated directly into sql, so they are
        # replaced to avoid a mismatch of number of format strings and parameters.
        params = [item for subparams in params for item in subparams if item is not TO_DELETE]
        return sql, params
```

While I was looking at this though I noticed that `TO_DELETE` isn't actually used anywhere. Is this left over from something you've changed since? If it's not needed then can you simplify the return of `resolve_combinable_parameter` such that you don't need to rezip into sql, params?

MatthewWilkes replied:
TIL you can have multiple `for .. in` constructs in a comprehension! You're right about TO_DELETE, too.

MatthewWilkes replied:
So, I had a play with `timeit` with some different implementations of this. The fastest one that passes tests involves returning a generator, but I'm too nervous to actually commit that.  

timeit table below, 1 million loops for a 6 parameter lookup and a mocked resolve_combinable_parameter, for reference:

For loop, no zip 5.960976593007217
Zip, sum, comprehension 6.835406245998456
Zip, double-comprehension 5.876567202998558
Zip, sum 5.648494772001868
Zip, chain 4.68141534399183
Zip, chain, tuple 5.421347933996003

jarshwah commented on django/db/models/sql/query.py:
```diff
@@ -988,6 +988,18 @@ def prepare_lookup_value(self, value, lookups, can_reuse, allow_joins=True):
+                        reuse=can_reuse,
+                        allow_joins=allow_joins
+                    ))
+                    used_joins |= set(k for k, v in self.alias_refcount.items() if v > pre_joins.get(k, 0))
```
This isn't my area of expertise. @akaariai have you got some time to look at this block in particular?

jarshwah commented on tests/expressions/tests.py:
```diff
@@ -391,6 +401,216 @@ def test_ticket_18375_chained_filters(self):
+            "name", "num_employees", "num_chairs"
+        ).order_by(
+            "name", "num_employees", "num_chairs"
+        )
```
Seems that this is unused?

jarshwah commented on tests/expressions/tests.py:
```diff
@@ -391,6 +401,216 @@ def test_ticket_18375_chained_filters(self):
+        )
+        for alias in queryset.query.alias_map.values():
+            if isinstance(alias, Join):
+                self.assertEqual(alias.join_type, constants.INNER)
```
Based on the conversation around https://github.com/django/django/pull/6216#discussion_r54401771 shouldn't these joins be LEFT since the start and end fields are nullable? The result is correct in this case but it wouldn't be if a query relied on the join being nullable. I'm struggling to come up with a query to exercise the problem. @akaariai you have anything for this?

MatthewWilkes replied:
Should this query not be equivalent to

    queryset = SimulationRun.objects.filter(midpoint__gte=F('start__time'), midpoint__lte=F('end__time'))

which also has `T3` and `expressions_time` as `INNER JOIN`?  In fact, their `alias_map`s are the same, modulo a different ordering of the join aliases.

I agree, I think this is wrong based on the comment from before, it's melting my mind a bit, but doesn't seem to differ from the current behaviour.

jarshwah replied:
Perhaps I misunderstood Anssi's original intent somehow, but the fact that your filter expression uses INNER joins seems like you're maintaining the status quo at the minimum. 

@charettes I've just pinged you on IRC but if you've got some thoughts on this I'd like to hear them.

charettes replied:
@jarshwah sorry for the delay.

I think we all agree that `(midpoint__gte=F('start__time'), midpoint__lte=F('end__time'))` should be equivalent to `(midpoint__range=(F('start__time'), F('end__time')))`.

I don't understand @akaariai intent on the previous comment as it looks like a correct optimization to use `INNER JOIN` in this case as rows would be filtered out by the `WHERE` clause anyway.

Maybe you could add tests to make sure `exclude(midpoint__range=(F('start__time'), F('end__time')))` uses `LEFT OUTER JOIN` and produces correct results like `(midpoint__gte=F('start__time'), midpoint__lte=F('end__time'))` does?

MatthewWilkes replied:
Thanks for this, I'll add the new test.

MatthewWilkes replied:
I've added this test and can confirm we see the same results and an LOUTER for both forms of the exclude query.

jarshwah commented:
I think that's about all you can do without Anssi clarifying previous comments. I'll try and give this a thorough review tomorrow for style, but the logic/tests/docs seem on point.

MatthewWilkes commented:
@jarshwah Thanks! I've just had a quick run over the diff and fixed a couple of minor style issues, hopefully we're in the home stretch now. Thanks again for your help up to this point, the PR is a lot stronger than it was back in February.

jarshwah commented on django/db/models/sql/query.py:
```diff
@@ -988,6 +988,18 @@ def prepare_lookup_value(self, value, lookups, can_reuse, allow_joins=True):
+                        self,
+                        reuse=can_reuse,
+                        allow_joins=allow_joins
+                    ))
```
I think this reads better as:

```
processed_values.append(
    sub_value.resolve_expression(self, reuse=can_reuse, allow_joins=allow_joins)
)
```

MatthewWilkes replied:
Aaand, done!

jarshwah commented:
That's it for me. If you fix up that one comment I made and then mark the ticket as ready for commit we can get a final review from someone other than me. Nice work here, thanks for persisting so long!

MatthewWilkes commented:
Great, I'll update the trac ticket and I'll re-trigger the oracle build (once everything else has passed, just in case) so that there's a full set of build statuses on this ticket for reference of whoever does the final review.

MatthewWilkes commented:
buildbot, test on oracle.

timgraham commented on django/contrib/postgres/fields/array.py:
```diff
@@ -84,7 +84,14 @@ def db_type(self, connection):
-            return [self.base_field.get_db_prep_value(i, connection, prepared=False) for i in value]
+            prepared_values = []
+            for inner_value in value:
+                if hasattr(inner_value, 'resolve_expression'):
```
Test are passing with this if statement removed, leaving only what's in the else branch. I guess this might be an attempt to solve that tests marked as expected failures. Is it worth keeping these modifications in light of the fact that they don't work or can a test be added to show their purpose?

MatthewWilkes replied:
I agree, this is a left-over part of that work. Having the code here doesn't change any error reporting, and it's relatively trivial, so I've removed it on the basis that stale code is bad.

timgraham commented on django/db/models/lookups.py:
```diff
@@ -190,6 +197,49 @@ class FieldGetDbPrepValueIterableMixin(FieldGetDbPrepValueMixin):
+            prepared_values.append(rhs_value)
+        return prepared_values
+
+    def process_rhs(self, compiler, connection):
```
Possible I missed something but I'm not seeing any tests failing with this method removed.

MatthewWilkes replied:
This is part of a mixin with only two uses in the standard codebase. Both overrode this method, but one of those wasn't necessary. I've removed that override and this method is now required for tests to pass.

I'm inclined to think it's better to have it here in a mixin rather than in the place I removed it from as it makes life easier for people implementing new lookups outside core.

timgraham commented on django/db/models/sql/query.py:
```diff
@@ -988,6 +988,16 @@ def prepare_lookup_value(self, value, lookups, can_reuse, allow_joins=True):
             pre_joins = self.alias_refcount.copy()
             value = value.resolve_expression(self, reuse=can_reuse, allow_joins=allow_joins)
             used_joins = [k for k, v in self.alias_refcount.items() if v > pre_joins.get(k, 0)]
+        elif isinstance(value, (list, tuple)):
```
Is this code self-explanatory enough not to need any comments? I have no clue...

MatthewWilkes replied:
I've added a couple, it never hurts to have more hints.

timgraham commented:
Cosmetic edits at https://github.com/MatthewWilkes/django/pull/1

MatthewWilkes commented:
@timgraham Merged your PR and addressed your comments. I squashed into one commit as I believe that's the policy. Let me know if you'd prefer your changes to go in independently and I'll isolate them.

timgraham commented:
merged in 4f138fe5a496a81115c4fba6615a517fc62c3b17, thanks!
